### DISCUSSION TITLE
how to know if connections is dead?

### DISCUSSION THREAD

**Discussion by bethebest0622** [curl/curl#12715] [Q&A] [answered]
### I did this

I pursuit the best performance, which means i want to reuse the connection, and want the connection never close.

I tried these options:

```
CURLSHcode r = curl_share_setopt(share_, CURLSHOPT_SHARE, CURL_LOCK_DATA_COOKIE);
  if (r) printf("Setting Shared Failed COOKIE: %s\n", curl_share_strerror(r));
  
  r = curl_share_setopt(share_, CURLSHOPT_SHARE, CURL_LOCK_DATA_DNS);
  if (r) EE("Setting Shared Failed DNS: %s\n", curl_share_strerror(r));

  r = curl_share_setopt(share_, CURLSHOPT_SHARE, CURL_LOCK_DATA_SSL_SESSION);
  if (r) printf("Setting Shared Failed SSL: %s\n", curl_share_strerror(r));

  r = curl_share_setopt(share_, CURLSHOPT_SHARE, CURL_LOCK_DATA_CONNECT);
  if (r) printf("Setting Shared Failed CONNECT: %s\n", curl_share_strerror(r));

  curl_easy_setopt(r, CURLOPT_WRITEFUNCTION, curl_cb);
  curl_easy_setopt(r, CURLOPT_SSL_VERIFYPEER, false);
  curl_easy_setopt(r, CURLOPT_ENCODING, "gzip");


  curl_easy_setopt(r, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_2_0);
  curl_easy_setopt(r, CURLOPT_DNS_CACHE_TIMEOUT, -1);

  curl_easy_setopt(r, CURLOPT_MAXCONNECTS, 128);

  // curl_easy_setopt(r, CURLOPT_VERBOSE, 1L);

  // curl_easy_setopt(r, CURLOPT_FORBID_REUSE, 0L);
  // curl_easy_setopt(r, CURLOPT_FRESH_CONNECT, 0L);

  curl_easy_setopt(r, CURLOPT_TCP_KEEPALIVE, 1L);
  curl_easy_setopt(r, CURLOPT_TCP_KEEPIDLE, 7200L);
  curl_easy_setopt(r, CURLOPT_MAXAGE_CONN, 7200L);
  curl_easy_setopt(r, CURLOPT_TCP_KEEPINTVL, 75L);

  // curl_easy_setopt(r, CURLOPT_TIMEOUT, 0);
  // curl_easy_setopt(r, CURLOPT_NOSIGNAL, 1L);

  curl_easy_setopt(r, CURLOPT_SHARE, share_);


```

### I expected the following

i think after i set these options, the CURL* wont build connections more than once(of course no resolving process). 

in the most request, it will ignore the build connection's process, as the PRETRANSFER_TIME is very short.

but there is always about 5% request rebuild the connection, i dont know why.

could you help on this? how can i check the connections cache in memory to know if the old connections dead or alive?

### curl/libcurl version

curl 8.5.0

### operating system

centos stream 8

**Comment by bagder** [comment, marked as answer]
Typically libcurl tells in the verbose output when it deems a connection needs to get closed.

If you have connections in the pool for a long time before trying to reuse them, `CURLOPT_MAXLIFETIME_CONN` will matter.

Some reflections on your code:

>  curl_easy_setopt(r, CURLOPT_ENCODING, "gzip");

1. This uses the old option name, it s called `CURLOPT_ACCEPT_ENCODING` since 7.21.6
2. We recommend setting it to `""` to let libcurl pick among the ones it supports

> curl_easy_setopt(r, CURLOPT_DNS_CACHE_TIMEOUT, -1);

To never timeout the DNS cache is aggressive, I hope you know what you are doing.

**Reply by bethebest0622** [reply]
`CURLOPT_MAXLIFETIME_CONN `, the default value is 0, if i don't revise it, it should be forever?

i want to describe what i am doing, it's very meaningful for me if you have more advice.

I have a very latency-sensitive code, need to post https request frequently. it wont run too long, but need fast.

so i try to reduce the latency as much as possible, so i set CURLOPT_DNS_CACHE_TIMEOUT -1, try my best to reuse the connection cache, make it open forever. but i still found some connections are rebuilding.

if you have more advice, will be of great help.

Thanks for your time

**Reply by bagder** [reply]
Oh sorry, I meant **`CURLOPT_MAXAGE_CONN`** and you already have that set.

**Reply by bagder** [reply]
> CURLOPT_DNS_CACHE_TIMEOUT -1, try my best to reuse the connection cache, make it open forever. but i still found some connections are rebuilding

The DNS cache will have zero effect on connection reuse since curl only resolves host names when it already decided that it cannot reuse a connection. But it will have an effect on name resolving.

**Reply by bethebest0622** [reply]
but i have set `CURLOPT_MAXAGE_CONN `as:

`curl_easy_setopt(r, CURLOPT_MAXAGE_CONN, 7200L);`

i thought the connection will last 2 hours at least, but seems not work.

could you let me know how to make all connections alive forever and get best speed? such as: ignore ssl, add host ip manually....

**Reply by bagder** [reply]
curl will reuse all the connections as long as they can be reused.

If you think you have connections that are not reused then the verbose output/tracing will help you understand why curl deems it can't reuse those connections.

**Reply by bethebest0622** [reply]
thanks.

I got the verbose:

* RESOLVE fapi.binance.com:443 is - old addresses discarded!

i think the problem is the host ip is not different, which means the fapi.binance.com map to multi ip address, which make the old connections rebuild. can i bind the fapi.binance.com to one fixed ip manually, let the discard address operation never happen?

**Reply by jay** [reply]
> * RESOLVE fapi.binance.com:443 is - old addresses discarded!

that message is only shown when [CURLOPT_RESOLVE](https://curl.se/libcurl/c/CURLOPT_RESOLVE.html) is used and an existing matching dns entry is discarded

https://github.com/curl/curl/blob/7161cb17c01dcff1dc5bf89a18437d9d729f1ecd/lib/hostip.c#L1249-L1267

**Comment by bethebest0622** [comment]
> > * RESOLVE fapi.binance.com:443 is - old addresses discarded!
> 
> that message is only shown when [CURLOPT_RESOLVE](https://curl.se/libcurl/c/CURLOPT_RESOLVE.html) is used and an existing matching dns entry is discarded
> 
> https://github.com/curl/curl/blob/7161cb17c01dcff1dc5bf89a18437d9d729f1ecd/lib/hostip.c#L1249-L1267
Hi Jay, could you provide some advice about how to maintain the connection forever.

from verbose, there is some more details:

```
* Connection 8 seems to be dead!
* Closing connection 8
* 17 bytes stray data read before trying h2 connection
* Found bundle for host fapi.binance.com: 0x241e4d0 [can multiplex]
* Hostname fapi.binance.com was found in DNS cache
*   Trying 54.238.73.188:443...
* Failed to set TCP_KEEPIDLE on fd 127
* Connected to fapi.binance.com (54.238.73.188) port 443 (#10)
```

is this mean tcp_keepidle failed to work?

**Reply by jay** [reply]
> curl_easy_setopt(r, CURLOPT_MAXAGE_CONN, 7200L);
> i thought the connection will last 2 hours at least, but seems not work.

...

> could you provide some advice about how to maintain the connection forever.

...

> * Failed to set TCP_KEEPIDLE on fd 127

Can you elaborate on why you want to maintain the connection forever? Just because you set it to two hours doesn't guarantee the connection is going to stay open that long, it depends on the network and server. The reason the default maximum age is 118 seconds is because the chances of successfully reusing a connection after that time drops dramatically.

Yes TCP_KEEPIDLE can help. I don't know why it fails to set for you since there is no error code. (See #12726 for proposed fix)

Curl currently [does not monitor connections in the connection pool](https://curl.se/docs/todo.html#Monitor_connections_in_the_conne). There is [curl_easy_upkeep](https://curl.se/libcurl/c/curl_easy_upkeep.html) for HTTP/2 PING but it does not work for shared connection caches, including the default shared connection cache in a multi. (See #12677 for proposed fix)

I would contact Binance and ask them what they recommend to keep the connection open. Maybe you can send some HEAD requests every so often. I connected to the server but it's geofenced and will not stay connected since I'm in the United States.

**Reply by bethebest0622** [reply]
the reason i need to maintain forever is: Binance is a trade platform, i need to send trade request by https. the request fast or not is very important for me.

So, I prefer to waste some resource to maintain the connection, so that can speed up(reuse the connection).

it's important for me, so i am very grateful that you can discuss with me.

the newest debug update from my side is:

the connection dead in lib/url.c line 810.
![image](https://github.com/curl/curl/assets/107900327/ac4746fc-156c-47e0-bf6f-9ae0d599f409)

here make dead TURE.

Thanks for your time.

**Reply by bethebest0622** [reply]
I guess Keep idle failed because I set keepidle as 720000，maybe too large. After I set it 7200,seems good

**Reply by jay** [reply]
Please explain in text and not screenshots. If curl is not reusing the connection due to pending input the question is what is the pending input. Normally for HTTP/2 connections like yours Curl_conn_is_alive curl should call [http2_connisalive](https://github.com/curl/curl/blob/7161cb17c01dcff1dc5bf89a18437d9d729f1ecd/lib/http2.c#L581-L624) which will do processing and then determine if the connection is "clean" or not. In other words, there may be some expected HTTP/2 data that is processed and that is fine. So unless there's a bug in this process we can assume there is some other data pending.

You can try [curl_global_trace](https://curl.se/libcurl/c/curl_global_trace.html)("all") after curl_global_init and set a debugfunction for the easy handle to show the verbose messages with OpenSSL trace messages included:

~~~c
/* Simple debug callback for CURLOPT_DEBUGFUNCTION.
 * The OpenSSL TLS trace messages are not shown when verbose unless a debug
 * callback is used. eg: * TLSv1.3 (IN), TLS handshake, Certificate (11):
 * https://github.com/curl/curl/blob/curl-8_5_0/lib/vtls/openssl.c#L3527-L3533
 */
int fdebug(CURL *handle, curl_infotype type,
           char *data, size_t size, void *userp)
{
  static const char s_infotype[CURLINFO_END][3] = {
    "* ", "< ", "> ", "{ ", "} ", "{ ", "} " };
  switch(type) {
  case CURLINFO_TEXT:
  case CURLINFO_HEADER_OUT:
  case CURLINFO_HEADER_IN:
    fprintf((FILE *)userp, "%s%.*s", s_infotype[type], (int)size, data);
    break;
  }
  return 0;
}

...

  curl_easy_setopt(curl, CURLOPT_DEBUGFUNCTION, fdebug);
  curl_easy_setopt(curl, CURLOPT_DEBUGDATA, (void *)stderr);
  curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
~~~

/cc @icing
